### ISSUE TITLE
Invert hue option.. Can it be brought back?

### ISSUE THREAD

**Issue by JCCR** [lswank/Tranquility#9]
Nocturne used to have this feature that not only inverted the screen, but it also let you invert the color hues. For those that didn't know about this: Blue colors became orange colors when the screen was inverted, but with the invert hue option everything stayed dark but orange became blue again for example.

Can this feature be implemented again for 10.9?

[![Bountysource](https://www.bountysource.com/badge/issue?issue_id=3167101)](https://www.bountysource.com/issues/3167101-invert-hue-option-can-it-be-brought-back?utm_source=3167101&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by ylluminate** [comment]
Seems that things have pretty much stagnated here.  Has anyone found another workable solution that this point?

**Comment by amitmarkel** [comment]
Well. Perhaps we eventually will have to dive into nocturne's code and understand exactly what it does, including what these deprecated APIs were, and then implement it with the current given set of API.

**Comment by dmitrym0** [comment]
Is nocturne source available somewhere? I would mind taking a look.

**Comment by ejwessel** [comment]
https://code.google.com/p/blacktree-nocturne/

**Comment by dmitrym0** [comment]
Thanks @ejwessel. I see now that Tranquility is Nocturne that compiles and with the hue option removed. I got Nocturne to compile but the code that actually applies the hue filtering fails with an error code that implies that filtering code has been removed completely in the latest OS X (private APIs and all). I'm sure someone has looked at this before... is there a thread anywhere that details what people have tried?

**Comment by amitmarkel** [comment]
I am not aware of a thread where people attempted learning other OSX API
calls, but you can see what we discussed in this thread.

On Feb 18, 2015, at 19:36, Dmitry <notifications@github.com> wrote:

Thanks @ejwessel <https://github.com/ejwessel>. I see now that Tranquility
is Nocturne that compiles and with the hue option removed. I got Nocturne
to compile but the code that actually applies the hue filtering fails with
an error code that implies that filtering code has been removed completely
in the latest OS X (private APIs and all). I'm sure someone has looked at
this before... is there a thread anywhere that details what people have
tried?

—
Reply to this email directly or view it on GitHub
<https://github.com/lswank/Tranquility/issues/9#issuecomment-74908923>.

**Comment by amitmarkel** [comment]
What I uploaded then: https://www.dropbox.com/sh/hf2tnaugskcnaey/AAACZ8sfC5IM_Y3-xXXROsA9a

**Comment by dmitrym0** [comment]
Looks like Nocturne creates a transparent window and applies a hue filter with an angle of Pi. The reason it doesn't work in latest OS X is because the private calls that were used in 10.4/10.5 are gone. The private calls were originally from a framework called CoreImage. Looking at the [documentation](https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html#//apple_ref/doc/filter/ci/CIHueAdjust) I see a very similar filter called [CIHueAdjust](https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CoreImageFilterReference/index.html#//apple_ref/doc/uid/TP30000136-SW42) that also takes an `angle` as a parameter. In principle, to fix this, all we have to do is reimplement the filter creation using the new API (which is public).

I'll do some further digging later.

**Comment by amitmarkel** [comment]
The details are correct; cool. I'm currently in a computer science exam period so I can't quite assist with the actual digging at this time. If this approach works then there will be no need for manipulating ICC profiles, which is a plus due to the different per application behavior.

**Comment by dmitrym0** [comment]
I spent some time looking at this issue. I'm not sure why `CGSNewCIFilterByName` has been removed, but it doesn't appear to have been replaced by anything public. `NSView` has a [backgroundFilters](https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CALayer_class/index.html#//apple_ref/occ/instp/CALayer/backgroundFilters) property which is expressly not supported for the root `CALayer` (see comment in the header). Thus you can add `CIFilter`s to a window but then they only affect the contents of said window (see a sample on [cocoasnippets](http://cocoasnippets.io/)). It seems like a potential approach would be to capture the screen, rasterize it into the window and then apply a filter to it. I can't see it working beyond a proof of concept though.

TLDR:

- Apple removed a private API call that let you add filters to the root layer of a window. There's no public API.
- Next step, proof of concept: take a screenshot, apply filters and see if it looks right.

**Comment by lswank** [comment]
Yes, `CGSNewCIFilterByName` returns `kCGErrorNotImplemented` in OS X 10.8+. Doing this on an `NSView` certainly won't work, but perhaps iterating over all of the `NSWindow` instances will. We've already seen from parts of the original code that the background of Finder is also an `NSWindow` instance.

There is a pretty clear (looking) way to handle applying a `CIFilter` to an `NSWindow` instance. The real questions are:
* Can we get all of the instances of `NSWindow`s in the world? (Likely yes.)
* Can we apply the same filter to them? (Likely yes, but we'll have to be clever.)
* Can that filter be the same filter we had before -- or something better? (Likely yes.)
* Is there an easier way to do this with color profiles? (From `class-dump`ing ColorSync.app, it doesn't seem like there's a really easy way to do this, but I am not sure.)
* Can we set an ICC color profile in code? (Probably yes, but I couldn't find quick documentation to do this, so it might not be all that common.)

"Now," you're probably asking yourself, "doesn't Apple do this with color inversion and color profiles anyway? Shouldn't it all be nicely documented?" It should, but the internal team doesn't always document like they should. If anyone is going to WWDC this year -- and yes, I know tickets aren't being sold quiet yet -- this is an excellent question for the CoreGraphics / CoreImage developers. At WWDC, the engineers always say a whole lot more than they should, because no one is watching over their shoulders. They also give you way more tools and way more hands-on time. So, if anyone wants to send me to WWDC, I will get any answers we can't. ;)

**Comment by lswank** [comment]
(Whoops. Didn't mean to close it.)

**Comment by msutherl** [comment]
🙏 Hoping somebody figures this out!

**Comment by jbenet** [comment]
@dmitrym0 any luck?

**Comment by dmitrym0** [comment]
I left it in late february and hasn't returned to it. @lswank had some [good idea](https://github.com/lswank/Tranquility/issues/9#issuecomment-76651304) that are worth trying though.

**Comment by lswank** [comment]
I punted on this for the time being. I'd love some ancillary assistance
rethinking the UI and cleaning the rest of the application up for the bits
that do work well before moving back to tackle this one. Anyone know any of
the CoreGraphics team guys? They could also provide some insight. (Or in a
different world, they might actually add in just the private method we need
for the new version of the operating system and tell us about it with a
wink-wink-nudge-nudge . . .)

**Comment by dmitrym0** [comment]
Looks like https://hazeover.com/ is doing something similar, maybe the author would be open to sharing his approach. Hazeover darkens everything but your front most window.

**Comment by amitmarkel** [comment]
Interesting; but it might be possible they're creating a transparent window behind the foremost window all the time, which would require no implementation for being interactive... Haven't tried the software though.

**Comment by pointum** [comment]
@dmitrym0 @amitmarkel HazeOver is a translucent black window that is reordered behind the front window

**Comment by reversefold** [comment]
I managed to get a display screenshot and invert its hues in a clone of Nocturne:
https://github.com/reversefold/blacktree-nocturne/commit/ad0f0873285fe49b0edaacb88ff880fb783df0b8

Lots of commented out code and possibly non-needed code as I was throwing everything at the wall to see what stuck. It's a promising line of research if nothing else. I plan to look into CGDisplayStream next as it appears to be the way to get a constantly updating view of the screen. Pair that with the click-through window I have working now and we may have a version of this which minimally works.

Apologies for any bad code, I'm not familiar with Apple's display frameworks.

**Comment by amitmarkel** [comment]
Great news; looking forward to seeing your approach working!
I hope OpenGL doesn't invoke the discrete video card, although it may work with gfxCardStatus to force the integrated Intel chip.

**Comment by fsargent** [comment]
Just put in another $10 on [Bountysource](https://www.bountysource.com/issues/3167101-invert-hue-option-can-it-be-brought-back). 
Really looking forward to this working.

**Comment by jbenet** [comment]
:( really wish this worked. just added $50 more to Bountysource.

**Comment by jbenet** [comment]
@reversefold wonder if your approach will be super fast or introduce display latency? any further luck getting it working?

**Comment by vn971** [comment]
Hi folks.
If anything, I found a way for it to work under Linux: https://gitlab.com/vn971/linux-color-inversion

**Comment by JCCR** [comment]
@vn971 nice work!

I didn't know about mlaily/NegativeScreen
Looks like Windows and Linux have this covered

**Comment by danielng01** [comment]
I managed to make it for 1 window, but there seems to be no way to get all NSWindow instances. As you can imagine this will be security issue to add CALayer to other process NSWindow.

Maybe with the Accessibility API, but again not possible. You can get CGWindowId, but I don't think this can be transformed to NSWindow and NSView.

The screenshot approach seems to be the best option which I will try next. I was trying to do this for 2 months so it's really not so easy.

I also wrote to one guy who is expert in ICC profiles and I wait for response. If ICC profile would work this will be best, but from my understanding this is not possible.

Is there some other way to hook to the graphics pipeline and apply color matrix before the image goes to the screen.

CGSAddWindowFilter seemed so nice solution. I really want it back :D 

![screen shot 2017-05-08 at 3 22 53 pm](https://cloud.githubusercontent.com/assets/1231553/25804205/3a5c1796-3403-11e7-94a4-a6f21646810c.png)
![screen shot 2017-05-08 at 4 23 39 pm](https://cloud.githubusercontent.com/assets/1231553/25804208/3a85b056-3403-11e7-992f-85f445472e42.png)
![screen shot 2017-05-08 at 4 25 15 pm](https://cloud.githubusercontent.com/assets/1231553/25804207/3a818b02-3403-11e7-8cc7-8e2157cc71b4.png)
![screen shot 2017-05-08 at 4 25 56 pm](https://cloud.githubusercontent.com/assets/1231553/25804209/3a86c310-3403-11e7-915d-c26bb33993d0.png)

**Comment by danielng01** [comment]
Update: I made it with really heavy use of C++ and with taking a screenshot of the desktop.

You can find working example here:
https://github.com/danielng01/Iris-Builds/raw/master/OSX/experiments/Iris-Smart-Inversion-OSX.zip
Demonstration video:
https://youtu.be/dQepAWXoVAY

Basically once upon a time Apple has this awesome private API `CGSNewCIFilterByName`, but they dropped it for some reason. I first tried with ICC profiles, but this is impossible since you can only change the gamma.

Then I tried making filter to `NSWindow`, but this only works for windows of the app and we don't have access to other program NSWindows. There is some way to get access with Apple Accessibility API, but the other app needs to whitelist you in the source code.

The only working approach for OSX Sierra and newer versions seems to be the screenshot approach. The best way was to try with the streaming API `CGDisplayStream`, but you can't ignore the window in which you are displaying the output and you get recursive output and no update of the screen. More info here:
http://stackoverflow.com/questions/18158390/avcapturescreeninput-and-kcgwindowlistoptiononscreenbelowwindow

Based on the Apple example here:
https://developer.apple.com/library/mac/samplecode/SonOfGrab/Introduction/Intro.html

I took a screenshot made a transparent window and put the screenshot to it. This thing updates every 20 ms for relatively good framerate and uses OpenGL rendering, to perform a little faster. After all the optimization it still uses 20% CPU on my Mac which is bad, but this seems to be the only approach. Even if you perform even more optimizations to drop 5% CPU it's still high.

You need to calculate the Smart inversion every frame and take a picture. Taking the picture alone tooks 10% and can't be optimized. I would welcome any other approaches, but it seems to me like you just can't do this faster on OSX anymore.

**Comment by JCCR** [comment]
@danielng01 
Hey, thanks for giving this a shot! Really want to see how it works, but whenever I run it I see the "temperature" of the colors go down (flux like effect) and then a few seconds later my screen turns black and I can only see the cursor. I am using nvidia's web drivers if that makes any difference.

**Comment by danielng01** [comment]
Maybe it's the drivers.

I just tested on Sierra and Yosemite both real Mac and Virtual machine and it works. The black thing is the overlay. Maybe the screenshot function fails on your machine. It's from the SonOfGrab demo see if you can get it working on your machine

Yes the temperature thing is normal ignore it.

**Comment by JCCR** [comment]
@danielng01 
Just tried it on the regular drivers, you are correct it works there.

**Comment by ylluminate** [comment]
@danielng01 same problem here.  Flux like effect (perhaps triggering Fl.ux) and then black main screen (5 monitors).  I'm using the native ATI drivers on 10.12.5ß5.

**Comment by danielng01** [comment]
How much CPU it uses. Is this usefull for me it's unusable on my old Mac.

@ylluminate I really don't know, I write Carbon and Cocoa for the first time. The function is `CGWindowListCreateImage`. Try with only 1 monitor connected.

You get f.lux effect, because I added the window to my software Iris (https://iristech.co/). I can't make working demo from scratch.

**Comment by JCCR** [comment]
@danielng01 
It works on my MBP from late-2013, Core i5 of the time. It's unusable for me :( but the effect is exactly what I hoped for.
I'm running macOS natively on another machine with better specs, Core i7 6700k but my GPU is an NVIDIA 900 series and it doesn't work with the native drivers. I have had issues with iBooks for example where it is just a black or sometimes transparent window.

**Comment by dmitrym0** [comment]
I get a black screen as well. Sierra, on an early 2013 Macbook Pro. Fans start as well, as soon as I start Iris.

**Comment by JCCR** [comment]
@danielng01 
I restarted my laptop so it would be at the lowest CPU usage possible, it looks like it's using 50-60% CPU (maybe a whole core since this CPU is dual core)

It's somewhat usable, not sure how fast it's actually updating but it looks like 10-15 FPS to me.
